allennlp/fairness/fairness_metrics.py

```diff
@@ -0,0 +1,466 @@
+except in degenerate cases.
+"""
+
+
+@Metric.register("independence")
+class Independence(Metric):
+    """
+    Independence. Assumes integer labels, with
+    each item to be classified having a single correct class.
+    """
+
+    def __call__(
+        self,
+        predicted_labels: torch.Tensor,
+        protected_variable_labels: torch.Tensor,
+        num_classes: Optional[int] = None,
+    ) -> Dict[int, torch.FloatTensor]:
+        """
+        # Parameters
+
+        predicted_labels : `torch.Tensor`, required.
+            A tensor of predicted integer class labels of shape (batch_size, ...). Represented as C.
+        protected_variable_labels : `torch.Tensor`, required.
+            A tensor of integer protected variable labels of shape (batch_size, ...). It must be the same
+            shape as the `predicted_labels` tensor. Represented as A.
+        num_classes : `int`, optional (default = `None`).
+            Number of classes. If not supplied, `num_classes` is inferred from `predicted_labels`.
+
+        # Returns
+
+        kl_divs : `Dict[int, torch.FloatTensor]`
+            A dictionary mapping each protected variable label a to the KL divergence of P(C | A = a) from P(C).
+        """
+        predicted_labels, protected_variable_labels = self.detach_tensors(
+            predicted_labels, protected_variable_labels
+        )
+
+        # Some sanity checks.
+        if predicted_labels.size() != protected_variable_labels.size():
+            raise ConfigurationError(
+                "protected_variable_labels must be of same size as predicted_labels but "
+                "found tensor of shape: {}".format(protected_variable_labels.size())
+            )
+        if num_classes is not None and (predicted_labels >= num_classes).any():
+            raise ConfigurationError(
+                "A predicted label contains an id >= {}, "
+                "the number of classes.".format(num_classes)
+            )
+
+        if num_classes is None:
+            num_classes = predicted_labels.max() + 1
+        num_protected_variables = protected_variable_labels.max() + 1
+
+        C_dist = Categorical(
+            predicted_labels.float().histc(bins=num_classes, min=0, max=num_classes - 1)
+            / predicted_labels.nelement()
+        )
+        kl_divs: Dict[int, torch.FloatTensor] = {}
+        # There currently does not exist a robust loopless way to compute the conditional distributions
+        # Assumes num_protected_variables is small
+        for a in range(num_protected_variables):
+            C_given_a_dist = Categorical(
+                predicted_labels[protected_variable_labels == a]
+                .float()
+                .histc(bins=num_classes, min=0, max=num_classes - 1)
+                / predicted_labels.nelement()
+            )
+            kl_divs[a] = kl_divergence(C_given_a_dist, C_dist)
```

Comment:
Also, this will compute the distributions correctly only for a single batch, right? We want that the distributions are computed based on all batches. (Again, you can take a look at other metrics for how this works). Ideally, the `__call__` function accumulates the distributions in some way, and the `get_metric()` function computes the final kl divergence. (I'm not sure how efficient this will be, though).